generate/src/contract.rs

```diff
@@ -0,0 +1,412 @@
+                Artifact {
+                    abi: artifact.abi.clone(),
+                    networks: artifact.networks.clone(),
+                    ..Artifact::empty()
+                }
+            };
+
+            DeployedFuture::from_args(web3, artifact)
+        }
+    }
+}
+
+fn expand_deploy(cx: &Context) -> Result<TokenStream> {
+    let ethcontract = &cx.runtime_crate;
+
+    // TODO(nlordell): not sure how contructor documentation get generated as I
+    //   can't seem to get truffle to output it
+    let doc = expand_doc("Generated by `ethcontract`");
+
+    let (input, arg) = match cx.artifact.abi.constructor() {
+        Some(contructor) => (
+            expand_inputs(cx, &contructor.inputs)?,
+            expand_inputs_call_arg(&contructor.inputs),
+        ),
+        None => (quote! {}, quote! {()}),
+    };
+
+    // TODO(nlordell): we don't handle duplicate library names
+    let libraries: Vec<_> = cx
+        .artifact
+        .bytecode
+        .undefined_libraries()
+        .map(|name| Param {
+            name: name.to_snake_case(),
+            kind: ParamType::Address,
+        })
+        .collect();
+    let lib_input = expand_inputs(cx, &libraries)?;
+
+    let link = if libraries.is_empty() {
+        let link_libraries = libraries.iter().map(|lib| {
+            let name = Literal::string(&lib.name);
+            let address = ident!(&lib.name);
+
+            quote! {
+                artifact.bytecode.link(#name, #address).expect("valid library");
+            }
+        });
+
+        quote! {
+            let mut artifact = artifact;
+            #( #link_libraries )*
+        }
+    } else {
+        quote! {}
+    };
+
+    Ok(quote! {
+        #doc
+        pub fn deploy<F, T>(
+            web3: &#ethcontract::web3::api::Web3<T> #input #lib_input ,
+        ) -> #ethcontract::DynDeployBuilder<Self>
+        where
+            F: #ethcontract::web3::futures::Future<Item = #ethcontract::json::Value, Error = #ethcontract::web3::Error> + Send + 'static,
+            T: #ethcontract::web3::Transport<Out = F> + 'static,
+        {
+            use #ethcontract::contract::DeployBuilder;
+            use #ethcontract::transport::DynTransport;
+            use #ethcontract::truffle::Artifact;
+            use #ethcontract::web3::api::Web3;
+
+            let transport = DynTransport::new(web3.transport().clone());
+            let web3 = Web3::new(transport);
+
+            let artifact = { // only clone the pieces we need
+                let artifact = Self::artifact();
+                Artifact {
+                    abi: artifact.abi.clone(),
+                    bytecode: artifact.bytecode.clone(),
+                    ..Artifact::empty()
+                }
+            };
+            #link
+
+            DeployBuilder::new(web3, artifact, #arg).expect("valid deployment args")
+        }
+    })
+}
+
+fn expand_function(cx: &Context, function: &Function) -> Result<TokenStream> {
```

Comment:
You are using Context now instead of just the needed Ident of the runtimeCrate. What was your thinking behind that?

Reply:
I need to pass the artifact as well, So It turned `ethcontract: &Ident, artifact: &Artifact` into `cx: &Context` in many places. Also, it now needs the `artifact_path` string literal as well because of the refactor.

Reply:
Also, we need to handle name collisions eventually, and this context struct will help pass that information around.